### ISSUE TITLE
error codes not returning under client.onerror and onreadychange

### ISSUE THREAD

**Issue by Band-Aid** [allenmichael/box-javascript-sdk#23]
Thanks for pushing out #21 
although the exception isn't being called in `client.onerror` and `client.onreadystatechange`

It looks like the above 2 isn't calling `handleErrors` inside.
Is there anyway to make it handle the error inside these two?

```
client.onerror = function () {
        // status returns 0
}
  
client.onreadystatechange = function () {
    // status returns 0
}
```

**Comment by allenmichael** [comment]
Sure thing, I'll take a look. Sorry for the delays. Is this when using `upload` or when using the `XMLHttpRequest` object in general?

**Comment by Band-Aid** [comment]
If we make this a general exception would that also cover uploads? Or are those different exceptions?

**Comment by Band-Aid** [comment]
We're good to handle upload error only :)
@allenmichael

**Comment by Band-Aid** [comment]
I was playing around with the code but figure out that this request may be difficult since when a 401 error is returned xmlhttprequest returns a status code of 0 (or whenever a request is incomplete) with no response.
For other errors (404, 429 etc) I was able to extract the header from the response.
Do you have any recommendations for getting the status code for incomplete requests?

**Comment by allenmichael** [comment]
Hey there! So, I had the same experience as you with the underlying `XMLHttpRequest` and even `fetch` functionality in browser. I could even surface 401 errors to JavaScript with calls to other Box endpoints. The only time I received a status code 0 is from the `upload` API. Since the `upload` API is a different domain (https://upload.box.com/api/2.0), I think there might be different CORs support for this domain. I was able to successfully handle a 409 conflict error on the `upload` API endpoint however. It seems like 401 is the only thing that triggers a status code 0.

It would be a question for the Box team ultimately.

Here's the notes I added on the PR I just merged: Fixed some issues with the HTTP client (both fetch and XMLHttpRequest) preventing 401 status surfacing. All non-200 error messages include a response object on the Error object that include status, statusText, headers (if retrievable), and data (if retrievable). The Upload endpoint doesn't support CORs and will not allow the browser to surface 401 errors and instead returns status of 0.
